### PR TITLE
ROC-3800 Response reminder emails

### PR DESCRIPTION
**Pull request by dharmendrak**
Jira link: https://tools.hmcts.net/jira/browse/ROC-3800

### CODE REVIEW

File: infrastructure/main.tf [hunk 1]
```diff
@@ -70,6 +70,18 @@ data "vault_generic_secret" "oauth_client_secret" {
   path = "secret/${var.vault_section}/ccidam/idam-api/oauth2/client-secrets/cmc-citizen"
 }
 
+module "quartz-store-db" {
+  source = "git@github.com:hmcts/moj-module-postgres?ref=cnp-449-tactical"
+  product = "quartz-postgres-db"
```
Comment (timja-kainos) [inline infrastructure/main.tf]: please user:
`${var.product}-scheduler`

don't need postgres-db in there, and rename the module to be scheduler

File: infrastructure/main.tf [hunk 2]
```diff
@@ -70,6 +70,18 @@ data "vault_generic_secret" "oauth_client_secret" {
...
+  location = "${var.location}"
+  env = "${var.env}"
+  postgresql_user = "quartzStore"
```
Comment (timja-kainos) [inline infrastructure/main.tf]: scheduler

File: infrastructure/main.tf [hunk 3]
```diff
@@ -70,6 +70,18 @@ data "vault_generic_secret" "oauth_client_secret" {
...
+  database_name = "quartzStore"
```
Comment (timja-kainos) [inline infrastructure/main.tf]: scheduler

File: job-scheduler/src/main/java/uk/gov/hmcts/cmc/scheduler/exceptions/JobException.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package uk.gov.hmcts.cmc.scheduler.exceptions;
+
+import uk.gov.hmcts.reform.logging.exception.AbstractLoggingException;
+
+import static uk.gov.hmcts.reform.logging.exception.AlertLevel.P1;
+import static uk.gov.hmcts.reform.logging.exception.ErrorCode.UNKNOWN;
+
+public class JobException extends AbstractLoggingException {
```
Comment (timja-kainos) [inline job-scheduler/src/main/java/uk/gov/hmcts/cmc/scheduler/exceptions/JobException.java]: don't extent AbstractLoggingException please, just do RuntimeException

File: job-scheduler/src/main/java/uk/gov/hmcts/cmc/scheduler/exceptions/JobNotFoundException.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package uk.gov.hmcts.cmc.scheduler.exceptions;
+
+import uk.gov.hmcts.reform.logging.exception.AbstractLoggingException;
+
+import static uk.gov.hmcts.reform.logging.exception.AlertLevel.P1;
+import static uk.gov.hmcts.reform.logging.exception.ErrorCode.UNKNOWN;
+
+public class JobNotFoundException extends AbstractLoggingException {
```
Comment (timja-kainos) [inline job-scheduler/src/main/java/uk/gov/hmcts/cmc/scheduler/exceptions/JobNotFoundException.java]: same here

File: job-scheduler/src/main/java/uk/gov/hmcts/cmc/scheduler/jobs/NotificationEmailJob.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+@Component
+public class NotificationEmailJob implements Job {
+
+    private static final Logger logger = LoggerFactory.getLogger(NotificationEmailJob.class);
+
+    @Autowired
+    private NotificationEmailService notificationEmailService;
+
+    @Override
+    public void execute(JobExecutionContext context) throws JobExecutionException {
```
Comment (timja-kainos) [inline job-scheduler/src/main/java/uk/gov/hmcts/cmc/scheduler/jobs/NotificationEmailJob.java]: this job should probably be removed in the initial PR, / add a job that only logs and doesn't send emails

File: src/main/java/uk/gov/hmcts/cmc/claimstore/events/claim/ClaimIssuedCitizenActionsHandler.java [hunk 7]
```diff
@@ -56,6 +69,37 @@ public void sendDefendantNotification(CitizenClaimIssuedEvent event) {
         }
     }
 
+    // This is added for spike, the actual implementation may vary based on the requirements in story
```
Comment (timja-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/events/claim/ClaimIssuedCitizenActionsHandler.java]: should this still be here?

File: infrastructure/main.tf [hunk 8]
```diff
@@ -70,6 +70,18 @@ data "vault_generic_secret" "oauth_client_secret" {
   path = "secret/${var.vault_section}/ccidam/idam-api/oauth2/client-secrets/cmc-citizen"
 }
 
+module "scheduler" {
+  source = "git@github.com:hmcts/moj-module-postgres?ref=cnp-449-tactical"
```
Comment (timja-kainos) [inline infrastructure/main.tf]: Please pass in a version parameter and set it to 10.3 it just got released a couple of days ago better to do now rather than later

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java [hunk 9]
```diff
@@ -186,11 +195,57 @@ private void linkToCase(User defendantUser, User anonymousCaseWorker, String let
...
+        String caseId,
+        String defendantId,
+        String defendantEmail,
+        CaseDetails caseDetails
+    ) {
+        Claim claim = ccdCaseDataToClaim.to(caseDetails.getId(), caseDetails.getData());
+        LocalDate responseDeadline = claim.getResponseDeadline();
+        String defendantName = claim.getClaimData().getDefendant().getName();
+        String claimantName = claim.getClaimData().getClaimant().getName();
+        ImmutableMap.Builder<String, Object> emailData = ImmutableMap.builder();
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java]: Please do the following instead:

```
ImmutableMap.Builder<String, Object> emailData = ImmutableMap.builder()
  .put("caseId", caseId)
  .put("caseReference", claim.getReferenceNumber())
```

Utilise benefits of builder fluent API.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java [hunk 10]
```diff
@@ -186,11 +195,57 @@ private void linkToCase(User defendantUser, User anonymousCaseWorker, String let
...
+        emailData.put("caseReference", claim.getReferenceNumber());
+        emailData.put("defendantEmail", defendantEmail);
+        emailData.put("defendantId", defendantId);
+        emailData.put("defendantName", defendantName);
+        emailData.put("claimantName", claimantName);
+        emailData.put("responseDeadline", responseDeadline);
+
+        jobService.scheduleJob(
+            JobData.builder()
+                .id(UUID.randomUUID().toString())
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java]: What about including some identifiable details in ID to make trouble shooting easier?

For example we could have `reminder:defence-due-in-5-days:000MC001-UUID.randomUUID().toString()` and `reminder:defence-due-in-1-day:000MC001-UUID.randomUUID().toString()`

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/ClaimRepository.java [hunk 11]
```diff
@@ -18,10 +18,10 @@
 public interface ClaimRepository {
 
     @SuppressWarnings("squid:S1214") // Pointless to create class for this
-    String SELECT_FROM_STATEMENT = "SELECT * FROM claim";
+        String SELECT_FROM_STATEMENT = "SELECT * FROM claim";
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/ClaimRepository.java]: Please revert here and line below.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+            throw new NotificationException(e);
+        }
+    }
+
+    @Recover
+    public void logNotificationFailure(
+        NotificationException exception,
+        Map<String, Object> emailData
+    ) {
+        String errorMessage = "Failure: "
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java]: Please use logger UI to interpolate error message.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+            + " failed to send response needed notification (" + emailData.get("caseReference")
+            + " to defendant at " + emailData.get("defendantEmail") + ") "
+            + " due to " + exception.getMessage();
+
+        logger.info(errorMessage, exception);
+    }
+
+    private Map<String, String> aggregateParams(Map<String, Object> emailData) {
+        ImmutableMap.Builder<String, String> parameters = new ImmutableMap.Builder<>();
+        parameters.put(CLAIM_REFERENCE_NUMBER, (String) emailData.get("caseReference"));
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java]: Use fluent API here as pointed earlier in other comment.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+        NotificationClient notificationClient,
+        NotificationsProperties notificationsProperties
+    ) {
+        this.notificationClient = notificationClient;
+        this.notificationsProperties = notificationsProperties;
+    }
+
+    @Retryable(value = NotificationException.class, backoff = @Backoff(delay = 200))
+    @Override
+    public void sendMail(Map<String, Object> emailData) {
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java]: Please use class instead of map for job details.

File: src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/functional/SubmitClaimTest.java [hunk 15]
```diff
@@ -35,9 +35,13 @@ public void before() {
 
     @Test
     public void shouldSuccessfullySubmitClaimDataAndReturnCreatedCase() {
+        UUID externalId = UUID.randomUUID();
         ClaimData claimData = testData.submittedByClaimantBuilder()
+            .withExternalId(externalId)
             .build();
 
+        commonOperations.submitPrePaymentClaim(externalId.toString(), claimant.getAuthorisation());
```
Comment (timja-kainos) [inline src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/functional/SubmitClaimTest.java]: this has nothing to do with your change

File: src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/BaseTest.java [hunk 16]
```diff
@@ -16,10 +16,7 @@
 @RunWith(SpringRunner.class)
 @SpringBootTest
 @TestPropertySource("/environment.properties")
-@ActiveProfiles({
-    "aat",
-    "mocked-database-tests"
-})
+@ActiveProfiles({"aat"})
```
Comment (timja-kainos) [inline src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/BaseTest.java]: is this change correct?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java [hunk 17]
```diff
@@ -196,11 +205,57 @@ private void linkToCase(User defendantUser, User anonymousCaseWorker, String let
...
+                .defendantId(defendantId)
+                .defendantEmail(defendantEmail)
+                .build(),
             CaseEvent.LINK_DEFENDANT
         );
+
+        scheduleEmailNotificationsForDefendantResponse(caseId, defendantId, defendantEmail, caseDetails);
+    }
+
+    private void scheduleEmailNotificationsForDefendantResponse(
```
Comment (timja-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java]: This needs to happen in the claim service,
this story is not only about CCD it should happen whether or not ccd is on

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/CoreCaseDataService.java [hunk 18]
```diff
@@ -105,7 +105,7 @@ public CaseReference savePrePayment(String externalId, String authorisation) {
 
             StartEventResponse startEventResponse = startCreate(authorisation, eventRequestData, user.isSolicitor());
 
-            CaseDataContent caseDataContent = CaseDataContent.builder()
+            CaseDataContent caseData = CaseDataContent.builder()
```
Comment (timja-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/CoreCaseDataService.java]: revert the changes in this file please, they are unnecessary and there's already lots of changes in this PR

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/NotificationEmailService.java [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+    private final ResponseNeededNotificationService responseNeededNotificationService;
+
+    public NotificationEmailService(ResponseNeededNotificationService responseNeededNotificationService) {
+        this.responseNeededNotificationService = responseNeededNotificationService;
+    }
+
+    @Retryable(
+        value = {JobExecutionException.class},
+        backoff = @Backoff(delay = 10000))
+    public void process(JobExecutionContext context) throws JobExecutionException {
```
Comment (timja-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/NotificationEmailService.java]: How many times is it retried? and if it fails after all those attempts what happens?
I can't see anything rescheduling the job

File: src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/BaseTest.java (hunk 16 above)
Comment (dharmendrak) [reply src/aat/java/uk/gov/hmcts/cmc/claimstore/tests/BaseTest.java]: need to revert

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/JobSchedulerService.java [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+
+    public void scheduleEmailNotificationsForDefendantResponse(
+        String defendantId,
+        String defendantEmail,
+        Claim claim
+    ) {
+        LocalDate responseDeadline = claim.getResponseDeadline();
+        String defendantName = claim.getClaimData().getDefendant().getName();
+        String claimantName = claim.getClaimData().getClaimant().getName();
+        ImmutableMap.Builder<String, Object> emailData = ImmutableMap.builder();
```
Comment (timja-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/JobSchedulerService.java]: build is please..!

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java [hunk 21]
```diff
@@ -56,21 +57,24 @@ public void sendMail(JobDetail jobDetail) {
...
-    ) {
+    ) throws JobExecutionException {
         JobDataMap emailData = jobDetail.getJobDataMap();
         logger.info("Failure: failed to send response needed notification ({} to defendant at {}) due to {}",
             emailData.get("caseReference"),
             emailData.get("defendantEmail"),
             exception.getMessage(),
             exception);
+
+        exception.setRefireImmediately(true);
```
Comment (timja-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java]: is refire immediately what we want? when does it stop? how many tries? what happens when it gives up?

Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java]: I am still testing this bit., pushed in to have your view

File: src/main/java/uk/gov/hmcts/cmc/claimstore/jobs/NotificationEmailJob.java [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(NotificationEmailJob.class);
+
+    @Autowired
+    public ResponseNeededNotificationService responseNeededNotificationService;
+
+    @Override
+    public void execute(JobExecutionContext context) throws JobExecutionException {
+        JobDetail jobDetail = context.getJobDetail();
+        responseNeededNotificationService.sendMail(jobDetail);
+        logger.info("Completed job work for id %s", jobDetail.getKey().getName());
```
Comment (timja-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/jobs/NotificationEmailJob.java]: debug? saying its complete doesn't give a lot of value

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    @Recover
+    public void logNotificationFailure(
+        NotificationException exception,
+        JobDetail jobDetail
+    ) {
+        JobDataMap emailData = jobDetail.getJobDataMap();
+        String caseReference = (String) emailData.get("caseReference");
+        logger.info("Failure: failed to send response needed notification ({} to defendant at {}) due to {}",
```
Comment (timja-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java]: warn maybe?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java [hunk 24]
```diff
@@ -37,29 +40,39 @@
...
-            notificationClient.sendEmail(
-                getEmailTemplates().getDefendantResponseNeeded(),
-                (String) emailData.get("defendantEmail"),
-                parameters,
-                (String) emailData.get("caseReference")
-            );
+            String caseReference = (String) emailData.get("caseReference");
+            Optional<Claim> claim = claimService.getClaimByReferenceAnonymous(caseReference);
+            if (claim.isPresent() && claim.get().getRespondedAt() == null) {
+                logger.debug("Claim is not yet responded for case {}, so sending notification to defendant",
```
Comment (timja-kainos) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java]: Do we need this?

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationServiceTest.java [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Test
+    public void sendShouldMailToDefendant() throws NotificationClientException {
+        //given
+        JobDetail jobDetail = mock(JobDetail.class);
+        String caseReference = "000MC004";
+        String defendantEmailAddress = "defendant@mail.com";
+
+        ImmutableMap<String, ? extends Serializable> data = ImmutableMap.of("caseReference", caseReference,
```
Comment (timja-kainos) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationServiceTest.java]: can you do `String, Object` instead? wildcard ? not needed here IMO

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/JobSchedulerService.java [hunk 26]
```diff
@@ -0,0 +1,123 @@
...
+        );
+
+        jobService.rescheduleJob(
+            getLastReminderJobData(claim, data),
+            responseDeadline.minusDays(lastReminderDay).atStartOfDay(ZoneOffset.UTC)
+        );
+
+    }
+
+    private JobData getLastReminderJobData(Claim claim, Map<String, Object> data) {
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/JobSchedulerService.java]: There is a lot of duplicates, please refactor it have one method `createReminderJobData(Claim claim, int numberOfDaysBeforeDeadline)`

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java [hunk 27]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    @Recover
+    public void logNotificationFailure(
+        NotificationException exception,
+        JobDetail jobDetail
+    ) {
+        JobDataMap emailData = jobDetail.getJobDataMap();
+        String caseReference = (String) emailData.get("caseReference");
+        logger.warn("Failure: failed to send response needed notification ({} to defendant at {}) due to {}",
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java]: I feel we should use ERROR severity here.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java [hunk 28]
```diff
@@ -0,0 +1,110 @@
...
+        this.notificationClient = notificationClient;
+        this.notificationsProperties = notificationsProperties;
+        this.appInsights = appInsights;
+        this.claimService = claimService;
+    }
+
+    @Retryable(value = NotificationException.class, backoff = @Backoff(delay = 200))
+    public void sendMail(JobDetail jobDetail) {
+        JobDataMap emailData = jobDetail.getJobDataMap();
+        Map<String, String> parameters = aggregateParams(emailData);
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java]: That logic should be invoked only when claim is eligible for response reminder.

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationServiceTest.java [hunk 29]
```diff
@@ -0,0 +1,76 @@
...
+@RunWith(MockitoJUnitRunner.class)
+public class ResponseNeededNotificationServiceTest extends BaseNotificationServiceTest {
+    public static final String EMAIL_TEMPLATE_ID = "email_template_id";
+    private ResponseNeededNotificationService responseNeededNotificationService;
+
+    @Mock
+    private AppInsights appInsights;
+    @Mock
+    private ClaimService claimService;
+
```
Comment (geek-soft) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationServiceTest.java]: Please remove blank line.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java [hunk 30]
```diff
@@ -0,0 +1,110 @@
...
+        AppInsights appInsights,
+        ClaimService claimService) {
+        this.notificationClient = notificationClient;
+        this.notificationsProperties = notificationsProperties;
+        this.appInsights = appInsights;
+        this.claimService = claimService;
+    }
+
+    @Retryable(value = NotificationException.class, backoff = @Backoff(delay = 200))
+    public void sendMail(JobDetail jobDetail) {
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java]: I feel more like `uk.gov.hmcts.cmc.claimstore.jobs.NotificationEmailJob#execute` should be responsible of checking whether notification can be send and `uk.gov.hmcts.cmc.claimstore.services.notifications.ResponseNeededNotificationService#sendMail` just send email.

@dharmendrak consider effort needed to change it. If too many I am happy to leave it as is for now. I know you have quite a few test cases that would need a change as well.

SUGESSTION ONLY

**Comment by pragnesh-patel** [comment]
some pending tests have not finished.
Tested locally
Checked table : qrtz_triggers for 2 entries in scheduler 

1. Check time in milliseconds (-1 day)
2. Check time in milleseconds (-5 days)

Request for more time - defendant journey

2 entries in scheduler as above, with adjusted jobs with respect to new 'Request for more time'

**TODO: check actual emails come through to defendant -1, -5 days before response deadline**

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java (hunk 30 above)
Comment (pragnesh-patel) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java]: how about a retest after changes @geek-soft - seems we have a preview env (from the view deployment link below)

Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java]: changes are cosmetic @pragnesh-patel

Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/notifications/ResponseNeededNotificationService.java]: pushed in, pull latest

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/JobSchedulerService.java [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+        );
+
+    }
+
+    private JobData createReminderJobData(Claim claim, Map<String, Object> data, int numberOfDaysBeforeDeadline) {
+        return JobData.builder()
+            .id("reminder:defence-due-in-"
+                + numberOfDaysBeforeDeadline
+                + "-day"
+                + (numberOfDaysBeforeDeadline > 1 ? "s" : "")
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/JobSchedulerService.java]: I did not even expect you to do it, I would be quite happy with `numberOfDaysBeforeDeadline + "-d" +` as it stands for days

Comment (geek-soft) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/JobSchedulerService.java]: Can you please use "String.format("reminder:defence-due-in-%d-d:%s-%s", ...)" here and few lines below as it make it easier to see the format.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/JobSchedulerService.java [hunk 32]
```diff
@@ -0,0 +1,106 @@
...
+                + " day"
+                + (numberOfDaysBeforeDeadline > 1 ? "s" : "")
+                + " before response deadline"
+            )
+            .jobClass(NotificationEmailJob.class)
+            .data(data).build();
+    }
+
+    private Map<String, Object> createNotificationData(String authorisation, String referenceNumber) {
+        User defendantUser = userService.getUser(authorisation);
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/JobSchedulerService.java]: Shouldn't we take it from the claim from `claim.defendantEmail` field? That would simplify code as we would not need to make IDAM calls and pass auth tokens.

**Comment by geek-soft** [comment]
@dharmendrak , @pragnesh-patel can you please retest scheduling to make sure last commit that changed the place we take defendant email address did not break the feature?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java [hunk 33]
```diff
@@ -218,10 +225,15 @@ private void revokeAccessToCase(User anonymousCaseWorker, String letterHolderId,
         );
     }
 
-    private void updateDefendantId(User defendantUser, String caseId, String defendantId) {
-        coreCaseDataService.update(
+    private CaseDetails updateDefendantId(User defendantUser, String caseId, String defendantId) {
+        String defendantEmail = defendantUser.getUserDetails().getEmail();
```
Comment (geek-soft) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/repositories/CCDCaseApi.java]: Please add `String defendantEmail` as a argument.

Then set that argument to `defendant@example.com` in uk.gov.hmcts.cmc.claimstore.repositories.support.CCDTestingSupportRepository#linkDefendantToClaim.

Please also rename `updateDefendantId` to `updateDefendantIdAndEmail` or similar.

**Comment by timja-kainos** [comment]
@dharmendrak says merge and the @pragnesh-patel is happy with it
